src/Nethermind/Nethermind.Network.Discovery/CompositeDiscoveryApp.cs

```diff
@@ -0,0 +1,219 @@
+    private readonly INetworkConfig _networkConfig;
+    private readonly IDiscoveryConfig _discoveryConfig;
+    private IConnectionsPool _connections { get; set; }
+
+    private IDiscoveryApp? _v4;
+    private IDiscoveryApp? _v5;
+
+    public CompositeDiscoveryApp(IApiWithNetwork api, ILogger logger, SameKeyGenerator privateKeyProvider)
+    {
+        _api = api;
+        _logger = logger;
+        _privateKeyProvider = privateKeyProvider;
+        _networkConfig = api.Config<INetworkConfig>();
+        _discoveryConfig = api.Config<IDiscoveryConfig>();
+        _connections = new DiscoveryConnectionsPool(_logger, _networkConfig, _discoveryConfig);
+    }
+
+    public event EventHandler<NodeEventArgs>? NodeAdded
+    {
+        add
+        {
+            if (_v4 != null) _v4.NodeAdded += value;
+            if (_v5 != null) _v5.NodeAdded += value;
+        }
+        remove
+        {
+            if (_v4 != null) _v4.NodeAdded -= value;
+            if (_v5 != null) _v5.NodeAdded -= value;
+        }
+    }
+
+    public event EventHandler<NodeEventArgs>? NodeRemoved
+    {
+        add
+        {
+            if (_v4 != null) _v4.NodeRemoved += value;
+            if (_v5 != null) _v5.NodeRemoved += value;
+        }
+        remove
+        {
+            if (_v4 != null) _v4.NodeRemoved -= value;
+            if (_v5 != null) _v5.NodeRemoved -= value;
+        }
+    }
+
+    public void Initialize(PublicKey masterPublicKey)
+    {
+        if ((_discoveryConfig.DiscoveryVersion & DiscoveryVersion.V4) != 0)
+            InitDiscoveryV4(_discoveryConfig, _privateKeyProvider);
+
+        if ((_discoveryConfig.DiscoveryVersion & DiscoveryVersion.V5) != 0)
+            InitDiscoveryV5(_privateKeyProvider);
+    }
+
+    public void InitializeChannel(IDatagramChannel channel)
+    {
+        _v4?.InitializeChannel(channel);
+        _v5?.InitializeChannel(channel);
+    }
+
+    public async Task StartAsync()
+    {
+        if (_v4 == null && _v5 == null) return;
+
+        Bootstrap bootstrap = new();
+        bootstrap.Group(new MultithreadEventLoopGroup(1));
+
+        if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            bootstrap.ChannelFactory(() => new SocketDatagramChannel(AddressFamily.InterNetwork));
+        else
+            bootstrap.Channel<SocketDatagramChannel>();
+
+        bootstrap.Handler(new ActionChannelInitializer<IDatagramChannel>(InitializeChannel));
+
+        await _connections.BindAsync(bootstrap, _networkConfig.DiscoveryPort);
+
+        await Task.WhenAll(
+            _v4?.StartAsync() ?? Task.CompletedTask,
+            _v5?.StartAsync() ?? Task.CompletedTask
+        );
+    }
+
+    public Task StopAsync() => Task.WhenAll(
+        _connections.StopAsync(),
+        _v4?.StopAsync() ?? Task.CompletedTask,
+        _v5?.StopAsync() ?? Task.CompletedTask
+    );
+
+    public void AddNodeToDiscovery(Node node)
+    {
+        // Add to v4 only
+        _v4?.AddNodeToDiscovery(node);
```

Comment:
Don't think v5 need to participate in this as of now, method implementation is actually empty there. Used by:
- `ProtocolsManager.AddNodeToDiscovery`
- `NodeSourceToDiscV4Feeder.AddToDiscoveryApp`

Reply:
Decided to add node to both v4 and v5. Will implement `AddNodeToDiscovery` in `DiscoveryV5App`.